src/t_stream.c

```diff
@@ -1386,6 +1401,174 @@ robj *createObjectFromStreamID(streamID *id) {
                         id->ms,id->seq));
 }
 
+/* Returns non-zero if the ID is 0-0. */
+int streamIDEqZero(streamID *id) {
+    return !(id->ms || id->seq);
+}
+
+/* A helper that returns non-zero if the range from 'start' to the stream's
+ * end doesn't contain a tombstone.
+ *
+ * NOTE: this assumes that the caller had verified that 'start' is less than
+ * 's->last_id'. */
+int streamRangeDoesNotContainTombstone(stream *s, streamID *start) {
+    streamID start_id;
+
+    if (!s->length || streamIDEqZero(&s->max_deleted_entry_id)) {
+        /* The stream is empty or has no tombstones. */
+        return 1;
+    }
+
+    /* Copy start ID, if given, or default to 0-0. */
+    if (start) {
+        start_id = *start;
+    } else {
+        start_id.ms = 0;
+        start_id.seq = 0;
+    }
+
+    if (streamCompareID(&s->first_id,&s->max_deleted_entry_id) > 0) {
+        /* The latest tombstone is before the first entry. */
+        return 1;
+    }
+
+    if (streamCompareID(&start_id,&s->max_deleted_entry_id) >= 0) {
+        /* The range doesn't include a tombstone. */
+        return 1;
+    }
+
+    /* The range includes a tombstone. */
+    return 0;
+}
+
+/* Replies with a consumerg group's current lag, that is the number of messages
+ * in the stream that are yet to be delivered. In case that the lag isn't
+ * available due to fragmentation, the reply to the client is a null. */
+void streamReplyWithCGLag(client *c, stream *s, streamCG *cg) {
+    int valid = 1;
+    uint64_t lag = 0;
+
+    if (!s->entries_added) {
+        /* The lag of a newly-initialized stream is 0. */
+        lag = 0;
+    } else if (cg->entries_read && streamRangeDoesNotContainTombstone(s,&cg->last_id)) {
```

Comment:
why do you check `cg->entries_read` here? seems like `if (streamRangeDoesNotContainTombstone(s,&cg->last_id))` is enough, no?

Reply:
maybe it's to ensure that cg->last_id is not 0-0..?

in which case maybe it's better to:
```
if (!s->entries_added) {
    /* The lag of a newly-initialized stream is 0. */
    lag = 0;
} else if (streamIDEqZero(&cg->last_id)) {
    serverAssert(cg->entries_read == 0);
    if (streamRangeDoesNotContainTombstone(s,NULL)) {
        /* The group is at 0-0 of a non-fragmented stream. */
        lag = s->length;
    } else {
        valid = 0;
    }
} else if (streamRangeDoesNotContainTombstone(s,&cg->last_id)) {
   serverAssert(cg->entries_read != 0);
   /* No fragmentation ahead means that the group's logical reads counter
     * is valid for performing the lag calculation. */
    lag = s->entries_added - cg->entries_read;
} else {
/* Attempt to retrieve the group's last ID logical read counter. */
    uint64_t entries_read = streamEstimateDistanceFromFirstEverEntry(s,&cg->last_id);
    if (entries_read) {
        /* A valid counter was obtained. */
        lag = s->entries_added - entries_read;
    } else {
        valid = 0;
    }
}